packages/vscode-extension/src/debug/taskTerminal/baseTunnelTaskTerminal.ts

```diff
@@ -0,0 +1,273 @@
+
+import { assembleError, err, FxError, ok, Result, UserError } from "@microsoft/teamsfx-api";
+import { envUtil, isV3Enabled } from "@microsoft/teamsfx-core";
+import { LocalTelemetryReporter } from "@microsoft/teamsfx-core/build/common/local";
+import { pathUtils } from "@microsoft/teamsfx-core/build/component/utils/pathUtils";
+
+import VsCodeLogInstance from "../../commonlib/log";
+import { ExtensionErrors, ExtensionSource } from "../../error";
+import * as globalVariables from "../../globalVariables";
+import { ProgressHandler } from "../../progressHandler";
+import {
+  TelemetryEvent,
+  TelemetryProperty,
+  TelemetrySuccess,
+} from "../../telemetry/extTelemetryEvents";
+import { getDefaultString, localize } from "../../utils/localizeUtils";
+import { Step } from "../commonUtils";
+import { ngrokTunnelDisplayMessages, TunnelDisplayMessages } from "../constants";
+import { doctorConstant } from "../depsChecker/doctorConstant";
+import { localTelemetryReporter } from "../localTelemetryReporter";
+import { BaseTaskTerminal } from "./baseTaskTerminal";
+
+export interface IBaseTunnelArgs {
+  type?: string;
+  env?: string;
+  output?: {
+    endpoint?: string;
+    domain?: string;
+  };
+}
+
+export const TunnelType = Object.freeze({
+  devTunnel: "dev-tunnel",
+  ngrok: "ngrok",
+});
+
+export type OutputInfo = {
+  file: string | undefined;
+  keys: string[];
+};
+
+export type EndpointInfo = {
+  src: string;
+  dist: string;
+};
+
+export abstract class BaseTunnelTaskTerminal extends BaseTaskTerminal {
+  protected static tunnelTaskTerminals: Map<string, BaseTunnelTaskTerminal> = new Map<
+    string,
+    BaseTunnelTaskTerminal
+  >();
+
+  protected outputMessageList: string[];
+
+  protected readonly progressHandler: ProgressHandler;
+  protected readonly step: Step;
+
+  constructor(taskDefinition: vscode.TaskDefinition, stepNumber: number) {
+    super(taskDefinition);
+
+    for (const task of BaseTunnelTaskTerminal.tunnelTaskTerminals.values()) {
+      task.close();
+    }
+
+    BaseTunnelTaskTerminal.tunnelTaskTerminals.set(this.taskTerminalId, this);
+
+    this.progressHandler = new ProgressHandler(
+      ngrokTunnelDisplayMessages.taskName,
+      stepNumber,
+      "terminal"
+    );
+    this.step = new Step(stepNumber);
+    this.outputMessageList = [];
+  }
+
+  public static async stopAll(): Promise<void> {
+    for (const task of BaseTunnelTaskTerminal.tunnelTaskTerminals.values()) {
+      task.close();
+    }
+  }
+
+  protected async resolveArgs(args: IBaseTunnelArgs): Promise<void> {
+    if (!args) {
+      throw BaseTaskTerminal.taskDefinitionError("args");
+    }
+
+    if (args.type) {
+      if (
+        typeof args.type !== "string" ||
+        !(Object.values(TunnelType) as string[]).includes(args.type)
+      ) {
+        throw BaseTaskTerminal.taskDefinitionError("type");
+      }
+    }
+
+    if (isV3Enabled()) {
+      if (typeof args.env !== "undefined" && typeof args.env !== "string") {
```

Comment:
Make `env` required in v3 to be consisitent with provision and deploy task?

Reply:
If `env` and `output` are not set, the tunnel serivce will still start, but the output will not be saved to the `.env` files.